File: cdm-sources/src/cdm/product/collateral/functions/CloneEligibleCollateralWithChangedTreatment.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.product.collateral.CollateralCriteria import CollateralCriteria
from cdm.product.collateral.EligibleCollateralSpecification import EligibleCollateralSpecification
from cdm.product.collateral.CollateralTreatment import CollateralTreatment

__all__ = ['CloneEligibleCollateralWithChangedTreatment']


@replaceable
def CloneEligibleCollateralWithChangedTreatment(inputSpecification: EligibleCollateralSpecification, changedCriteria: CollateralCriteria, changedTreatment: CollateralTreatment) -> EligibleCollateralSpecification:
    """
    Creates a new Eligible Collateral Specification based on an input specification but with one changed criteria with a changed treatment.
    
    Parameters 
    ----------
    inputSpecification : EligibleCollateralSpecification
    
    changedCriteria : CollateralCriteria
    
    changedTreatment : CollateralTreatment
    
    Returns
    -------
    outputSpecification : EligibleCollateralSpecification
    
    """
    _pre_registry = {}
    self = inspect.currentframe()
    
    # conditions
    
    @rosetta_local_condition(_pre_registry)
    def condition_0_NoLogicApplied(self):
        return (((not rosetta_attr_exists(rosetta_resolve_attr(rosetta_resolve_attr(self, "changedCriteria"), "AllCriteria"))) and (not rosetta_attr_exists(rosetta_resolve_attr(rosetta_resolve_attr(self, "changedCriteria"), "AnyCriteria")))) and (not rosetta_attr_exists(rosetta_resolve_attr(rosetta_resolve_attr(self, "changedCriteria"), "NegativeCriteria"))))
    # Execute all registered conditions
    execute_local_conditions(_pre_registry, 'Pre-condition')
    
    outputSpecification =  rosetta_resolve_attr(self, "inputSpecification")
    outputSpecification = _get_rosetta_object('EligibleCollateralSpecification', 'criteria', _get_rosetta_object('EligibleCollateralCriteria', 'treatment', rosetta_resolve_attr(self, "changedTreatment")))
    
    
    return outputSpecification

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)

File: cdm-sources/cdm/product/collateral/functions/CloneEligibleCollateralWithChangedTreatment.java
package cdm.product.collateral.functions;

import cdm.product.collateral.AllCriteria;
import cdm.product.collateral.AnyCriteria;
import cdm.product.collateral.CollateralCriteria;
import cdm.product.collateral.CollateralTreatment;
import cdm.product.collateral.EligibleCollateralSpecification;
import cdm.product.collateral.EligibleCollateralSpecification.EligibleCollateralSpecificationBuilder;
import cdm.product.collateral.NegativeCriteria;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.functions.ConditionValidator;
import com.rosetta.model.lib.functions.ModelObjectValidator;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperS;
import java.util.Optional;
import javax.inject.Inject;

import static com.rosetta.model.lib.expression.ExpressionOperators.*;

@ImplementedBy(CloneEligibleCollateralWithChangedTreatment.CloneEligibleCollateralWithChangedTreatmentDefault.class)
public abstract class CloneEligibleCollateralWithChangedTreatment implements RosettaFunction {
	
	@Inject protected ConditionValidator conditionValidator;
	
	@Inject protected ModelObjectValidator objectValidator;

	/**
	* @param inputSpecification 
	* @param changedCriteria 
	* @param changedTreatment 
	* @return outputSpecification 
	*/
	public EligibleCollateralSpecification evaluate(EligibleCollateralSpecification inputSpecification, CollateralCriteria changedCriteria, CollateralTreatment changedTreatment) {
		// pre-conditions
		conditionValidator.validate(() -> notExists(MapperS.of(changedCriteria).<AllCriteria>map("getAllCriteria", collateralCriteria -> collateralCriteria.getAllCriteria())).and(notExists(MapperS.of(changedCriteria).<AnyCriteria>map("getAnyCriteria", collateralCriteria -> collateralCriteria.getAnyCriteria()))).and(notExists(MapperS.of(changedCriteria).<NegativeCriteria>map("getNegativeCriteria", collateralCriteria -> collateralCriteria.getNegativeCriteria()))),
			"");
		
		EligibleCollateralSpecification.EligibleCollateralSpecificationBuilder outputSpecificationBuilder = doEvaluate(inputSpecification, changedCriteria, changedTreatment);
		
		final EligibleCollateralSpecification outputSpecification;
		if (outputSpecificationBuilder == null) {
			outputSpecification = null;
		} else {
			outputSpecification = outputSpecificationBuilder.build();
			objectValidator.validate(EligibleCollateralSpecification.class, outputSpecification);
		}
		
		return outputSpecification;
	}

	protected abstract EligibleCollateralSpecification.EligibleCollateralSpecificationBuilder doEvaluate(EligibleCollateralSpecification inputSpecification, CollateralCriteria changedCriteria, CollateralTreatment changedTreatment);

	public static class CloneEligibleCollateralWithChangedTreatmentDefault extends CloneEligibleCollateralWithChangedTreatment {
		@Override
		protected EligibleCollateralSpecification.EligibleCollateralSpecificationBuilder doEvaluate(EligibleCollateralSpecification inputSpecification, CollateralCriteria changedCriteria, CollateralTreatment changedTreatment) {
			EligibleCollateralSpecification.EligibleCollateralSpecificationBuilder outputSpecification = EligibleCollateralSpecification.builder();
			return assignOutput(outputSpecification, inputSpecification, changedCriteria, changedTreatment);
		}
		
		protected EligibleCollateralSpecification.EligibleCollateralSpecificationBuilder assignOutput(EligibleCollateralSpecification.EligibleCollateralSpecificationBuilder outputSpecification, EligibleCollateralSpecification inputSpecification, CollateralCriteria changedCriteria, CollateralTreatment changedTreatment) {
			outputSpecification = toBuilder(inputSpecification);
			
			outputSpecification
				.getOrCreateCriteria(0)
				.setTreatment(changedTreatment);
			
			return Optional.ofNullable(outputSpecification)
				.map(o -> o.prune())
				.orElse(null);
		}
	}
}
